### PR TITLE
new assertions

### PR DESCRIPTION
**Pull request by deltics**

### CODE REVIEW

File: matchers/maps/utils.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+	cmp := compareFuncFor[K, V](opts...)
+	for k, v := range m {
+		if vGot, ok := c[k]; !ok || !cmp(v, vGot) {
+			return false
+		}
+	}
+
+	return true
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/maps/utils.go:70]: yes, the logic was reversed but the call to the function also reversed the args; I have fixed the call sites

File: matchers/slices/appendToReport.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+type JustifyPrefix bool
+
+func AppendToReport(r []string, s any, pfx string, opts ...any) []string {
+	v := reflect.ValueOf(s)
+	switch {
+	case v.IsNil():
+		return append(r, pfx+" nil")
+	case v.Len() == 0:
+		return append(r, pfx+" <empty slice>")
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/slices/appendToReport.go:20]: thanks!

File: ordered_test.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+				Expect(2).To(BeGreaterThan(1))
+			},
+		},
+		{Scenario: "expected not greater than and was equal",
+			Act: func() {
+				Expect(2).ToNot(BeGreaterThan(2))
+			},
+		},
+	})
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply ordered_test.go:20]: edge cases are covered in the tests for the matcher itself; these tests provide coverage for the factory functions

File: ordered_test.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+				Expect(1).To(BeLessThan(2))
+			},
+		},
+		{Scenario: "expected not less than and was equal",
+			Act: func() {
+				Expect(2).ToNot(BeLessThan(2))
+			},
+		},
+	})
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply ordered_test.go:37]: matcher edge cases are covered in the matcher tests; these tests cover the factory functions

File: parallel.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+// The TestingG must be a *testing.T; any other implementation will return
+// false.
+//
+// For *testing.T values, reflection is used to examine the internal state
+// to determine if it is running in parallel.
+func isParallel(t TestingT) bool {
+	if _, ok := T().(*testing.T); !ok {
+		// tests cannot be parallel if the TestingT is not a *testing.T
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply parallel.go:20]: as noted in the comment, if the T() is not a *testing.T then it is not possible that the parameter t is parallel, so the more complex checks on t itself are not necessary in that case

File: matchers/slices/slice.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		for i, target := range cpy {
+			if cmp(item, target) {
+				cpy = append(cpy[:i], cpy[i+1:]...)
+				break
+			}
+		}
+	}
+
+	return len(cpy) == 0
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/slices/slice.go:30]: new test case added

File: matchers/strings/contains.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+	return strings.Contains(string(got), string(m.Expected))
+}
+
+func (m ContainsMatch) OnTestFailure(got string, opts ...any) []string {
+	if opt.IsSet(opts, opt.ToNotMatch(true)) {
+		return []string{
+			"expected: string not containing: " + opt.ValueAsString(m.Expected, opts...),
+			"got     : " + opt.ValueAsString(got, opts...),
+			fmt.Sprintf("             %s", strings.Repeat("^", len(m.Expected))),
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/strings/contains.go:27]: I udnerstand the issue; for most cases this is unlikely to be a problem and the caret will be helpful in the majority of cases.  This will be dealt with as an issue in the future, if it proves more problematic than expected

File: expect.go [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+	if matcher.Match(e.subject, opts...) {
+		switch special := matcher.(type) {
+		case interface{ MatchAny(any, ...any) bool }:
+			if special.MatchAny(e.subject, opts...) {
+				return
+			}
+		default:
+			return
+		}
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply expect.go:294]: I like it!  Thanks.

File: expect_isEmpty.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+	}
+	if result.method == "" {
+		result = hasLength[uint](e.subject)
+	}
+	if result.method == "" {
+		result = hasLength[int64](e.subject)
+	}
+	if result.method == "" {
+		result = hasLength[uint64](e.subject)
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply expect_isEmpty.go:143]: Good idea, thanks!

File: expect_test.go [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+
+type heterogenousMatcher[T any] struct{ result bool }
+
+func (m heterogenousMatcher[T]) Match(T, ...any) bool {
+	return true
+}
+
+func (m heterogenousMatcher[T]) MatchAny(got any, _ ...any) bool {
+	return m.result
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply expect_test.go:601]: The result from `Match()` in a heterogenous matcher is not significant since the `MatchAny()` method will be used in preference when present.  By convention, heterogenous matchers always return `true` from `Match()`, signalling compatibility with the typed parameter `T`.

File: matchers/contexts/valueMatcher.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	Key      K
+	Expected V
+}
+
+func (vm *ValueMatcher[K, V]) Match(_ context.Context, _ ...any) bool {
+	return true
+}
+
+func (vm ValueMatcher[K, V]) MatchAny(got any, opts ...any) bool {
+	ctx := got.(context.Context)
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/contexts/valueMatcher.go:21]: In a heterogenous matcher the type of `got` in a `MatchAny` implementation is guaranteed to be of type `T`.  Having reviewed this however, it appears that the context matchers do not need to be heterogenous matchers at all!
